sympy/solvers/tests/test_solvers.py

```diff
@@ -1372,12 +1372,16 @@ def test_issue_5901():
     assert solve([f(x) - 3*f(x).diff(x)], f(x)) == \
         {f(x): 3*D}
     assert solve([f(x) - 3*f(x).diff(x), f(x)**2 - y + 4], f(x), y) == \
-        [{f(x): 3*D, y: 9*D**2 + 4}]
-    assert solve(-f(a)**2*g(a)**2 + f(a)**2*h(a)**2 + g(a).diff(a),
-                h(a), g(a), set=True) == \
-        ([g(a)], {
-        (-sqrt(h(a)**2*f(a)**2 + G)/f(a),),
-        (sqrt(h(a)**2*f(a)**2+ G)/f(a),)})
+        [(3*D, 9*D**2 + 4)]
+    eq = -f(a)**2*g(a)**2 + f(a)**2*h(a)**2 + g(a).diff(a)
+    syms = [g(a), h(a)]
+    syms_found, sols_found = solve(eq, h(a), g(a), set=True)
+    assert syms_found == syms
+    # For these expressions dummy_eq doesn't work so we extract the dummy
+    X1 = list(sols_found)[0][0]
+    sols = { (g(a), +sqrt(X1**2*f(a)**2 - G)/f(a)),
+             (g(a), -sqrt(X1**2*f(a)**2 - G)/f(a)) }
+    assert sols == sols_found
```

Comment:
I'm still unsure why this is being changed. If the behaviour of solve has not changed then why is the test changed?

Reply:
> I'm still unsure why this is being changed. If the behaviour of solve has not changed then why is the test changed?

`solve()` is returning the correct solution in its own way. We are solving for both `g(a)` and `h(a)`. And we get this as the solution 
`([g(a), h(a)], {(g(a), sqrt(f(a)**2*g(a)**2 - Derivative(g(a), a))/f(a)), (g(a), -sqrt(f(a)**2*g(a)**2 - Derivative(g(a), a))/f(a))}`

Previously, `solve()` was returning only `g(a)`. So, in my opinion, the fix you made is working correctly.

However, we need to somewhat standardize the behaviour of solve. Unlike this example where solve is returning solutions separately for both variables, this is not the case with some other examples as I mentioned in [this comment](https://github.com/sympy/sympy/pull/20790#discussion_r580150396).

Reply:
Okay, I see now that the new result is correct. Which part of the changes in this PR fixes that?

I think the test can still be written more like the way it was though. The `X` here was introduced because of a Dummy that is no longer there I think.

Reply:
>  Which part of the changes in this PR fixes that?

That I haven't been able to figure out yet. From what I've understood till now is that different processed equations are passed to `_solve()` for this PR and master. Maybe that's what is causing the changes.

For master, `_solve()` gets this `[_Dummy_29 - _X0**2*f(a)**2 + _X1**2*f(a)**2]` whereas for the PR, `_solve()` gets this `[_Dummy_29 + _X0**2*f(a)**2 - _X1**2*f(a)**2]`. 

> I think the test can still be written more like the way it was though.

Yes, agreed.

Reply:
More importantly, the current fix isn't giving the correct answer for this case. 
`solve(-f(a)**2*g(a)**2 + f(a)**2*h(a)**2 + g(a).diff(a), h(a), g(a), set=True)` should return 

`([g(a)], {(sqrt(f(a)**2*h(a)**2 - Derivative(g(a), a))/f(a),), (-sqrt(f(a)**2*h(a)**2 - Derivative(g(a), a))/f(a),)})`.

However, there's a positive sign instead of a negative sign before Derivative... in the current PR. This missed my eye previously.

Reply:
> For master, `_solve()` gets this `[_Dummy_29 - _X0**2*f(a)**2 + _X1**2*f(a)**2]` whereas for the PR, `_solve()` gets this `[_Dummy_29 + _X0**2*f(a)**2 - _X1**2*f(a)**2]`.

That suggests a bug somewhere...

Reply:
That bug has is fixed now in the new changes. Is there anything more that needs to be taken care of?

Reply:
What was the cause of the bug?

Reply:
Let me check in master

Reply:
The bug is introduced in [this line](https://github.com/sympy/sympy/blob/5fad778de7796871b38ed9c4e71ca62671d857cf/sympy/solvers/solvers.py#L992) by `recast_to_symbol()`

Reply:
For both the cases `recast_to_symbol` is getting the same processed expression(`[-f(a)**2*g(a)**2 + f(a)**2*h(a)**2 + Derivative(g(a), a)]`) as `f` but different arguments in place of `symbols`.

In master, the `symbols` is `[g(a), h(a)]` whereas in the PR it is the same list but reversed `[h(a), g(a)]`.

I must admit though, I'm not sure what `recast_to_symbol()` even does.

Reply:
I think `recast_to_symbols` replaces non-symbol unknowns with symbols so e.g. if if `f(x)` is one of the unknowns then `recast_to_symbols` will replace `f(x)` with `_X0` so that it is a plain symbol in the rest of the `solve` code.

Reply:
I see, maybe `recast_to_symbol()` shouldn't be dependent on the order of symbols?

Reply:
I guess so. How does it depend on the order of the symbols?

Reply:
See [this](https://live.sympy.org/?evaluate=a%3DSymbol(%27a%27)%0A%23--%0Af%2C%20g%2C%20h%20%3D%20map(Function%2C%20%27fgh%27)%0A%23--%0Afrom%20sympy.solvers.solvers%20import%20recast_to_symbols%0A%23--%0A%20eq%3D%5B-f(a)**2*g(a)**2%20%2B%20f(a)**2*h(a)**2%20%2B%20Derivative(g(a)%2C%20a)%5D%0A%23--%0Arecast_to_symbols(eq%2C%20%5Bg(a)%2Ch(a)%5D)%0A%23--%0Arecast_to_symbols(eq%2C%20%5Bh(a)%2Cg(a)%5D)%0A%23--%0A)

Reply:
Oh, that isn't a bug. The output of `recast_to_symbols` is equivalent. It just has `X0` swapped with `X1`.

Reply:
Yeah you are right, I didn't notice that.

Reply:
I guess then we shouldn't worry too much about this since the solve output is correct now.